### PR TITLE
Make font size configurable

### PR DESCRIPTION
**Pull request by ramen**
Allow user to change the font size and persist this setting.

- Enable Settings plugin
- Move Logging to Settings
- Add ReaSpeechNumberInput
- Enable tab selection for ReaSpeechTabBar

### CODE REVIEW

**Review by mikeylove — APPROVED** [review]
solid, works great! i see what you mean with the -/+ buttons next to the input - it'd be nice if the buttons could respond to the boundary states (disabling when you can't go any further), but nothing stood out in the ImGui docs for doing that. 🤔

i also pulled this locally to play around with the tab bar and am pretty puzzled by that behavior. i noticed that we're not using `Storage` state on that so tried to just pop in a little memory cell but didn't really get anywhere quickly. curious! 🤔🤔🤔🤔🤔🤔🤔

**Comment by mikeylove** [comment]
playing around with the tab bar again this morning and realizing that i forgot to make a note that the only visual weirdness i ended up seeing was when the font size is cranked up high and the text is clipped on the "Choose file" button for the file selector. i don't consider this a blocker for this particular pr. :)
 
Screenshot 2024-12-19 at 10 35 56

**Comment by mikeylove** [comment]
after digging in a bit (really wanted to understand what was going on, still not 100% there), i ended up with something that appears to be functional. basically,
- reverting the changes to the tab bar
- removing the `ReaSpeechUI:refresh()` function and the `font_size.on_set` declaration that invoked it
- adding `Fonts:check(ctx)` (invoked in the main loop just after assigning `ctx`), which checks the current font size value against the last loaded one and does a detach/attach for the two font objects

```
diff --git a/reascripts/ReaSpeech/source/Fonts.lua b/reascripts/ReaSpeech/source/Fonts.lua
index 3e14b9a..9a1f6d2 100644
--- a/reascripts/ReaSpeech/source/Fonts.lua
+++ b/reascripts/ReaSpeech/source/Fonts.lua
@@ -29,10 +29,36 @@ function Fonts:init(ctx)
     end
   }
 
-  self.main = ImGui.CreateFont('sans-serif', self.size:get())
+  self.last_font_size = self.size:get()
+
+  self:load_and_attach(ctx, self.last_font_size)
+end
+
+function Fonts:check(ctx)
+  local current_font_size = self.size:get()
+
+  if current_font_size == self.last_font_size then
+    return
+  end
+
+  self.last_font_size = current_font_size
+
+  self:load_and_attach(ctx, current_font_size)
+end
+
+function Fonts:load_and_attach(ctx, font_size)
+  if ImGui.ValidatePtr(self.main, 'ImGui_Font*') then
+    ImGui.Detach(ctx, self.main)
+  end
+
+  self.main = ImGui.CreateFont('sans-serif', font_size)
   ImGui.Attach(ctx, self.main)
 
-  self.bold = ImGui.CreateFont('sans-serif', self.size:get(), ImGui.FontFlags_Bold())
+  if ImGui.ValidatePtr(self.bold, 'ImGui_Font*') then
+    ImGui.Detach(ctx, self.bold)
+  end
+
+  self.bold = ImGui.CreateFont('sans-serif', font_size, ImGui.FontFlags_Bold())
   ImGui.Attach(ctx, self.bold)
 end
 
diff --git a/reascripts/ReaSpeech/source/ReaSpeechMain.lua b/reascripts/ReaSpeech/source/ReaSpeechMain.lua
index 4304464..410d164 100644
--- a/reascripts/ReaSpeech/source/ReaSpeechMain.lua
+++ b/reascripts/ReaSpeech/source/ReaSpeechMain.lua
@@ -27,6 +27,7 @@ function ReaSpeechMain:loop()
   return function()
     if app:presenting() or app:is_open() then
       ctx = Ctx()
+      Fonts:check(ctx)
       Trap(function() app:react() end)
       reaper.defer(self:loop())
     end
diff --git a/reascripts/ReaSpeech/source/ReaSpeechUI.lua b/reascripts/ReaSpeech/source/ReaSpeechUI.lua
index dae08ab..7f5da92 100644
--- a/reascripts/ReaSpeech/source/ReaSpeechUI.lua
+++ b/reascripts/ReaSpeech/source/ReaSpeechUI.lua
@@ -78,12 +78,6 @@ ReaSpeechUI.config_flags = function ()
   return ImGui.ConfigFlags_DockingEnable()
 end
 
-function ReaSpeechUI:refresh()
-  Ctx.ctx = nil
-  local tab_bar = self.controls_ui.tab_bar
-  tab_bar:set(tab_bar:value())
-end
-
 function ReaSpeechUI:react()
   for _, handler in pairs(self.react_handlers) do
     Trap(handler)
diff --git a/reascripts/ReaSpeech/source/ReaSpeechWidgets.lua b/reascripts/ReaSpeech/source/ReaSpeechWidgets.lua
index 81f1783..7e826dd 100644
--- a/reascripts/ReaSpeech/source/ReaSpeechWidgets.lua
+++ b/reascripts/ReaSpeech/source/ReaSpeechWidgets.lua
@@ -254,25 +254,15 @@ ReaSpeechTabBar.new = function (options)
     options = options,
   })
 
-  o._set = o.set
-  function o:set(key)
-    o._key_to_select = key
-  end
-
   return o
 end
 
 ReaSpeechTabBar.renderer = function (self)
   if ImGui.BeginTabBar(ctx, 'TabBar') then
     for _, tab in pairs(self.options.tabs) do
-      local item_flags = 0
-      if self._key_to_select and tab.key == self._key_to_select then
-        item_flags = ImGui.TabItemFlags_SetSelected()
-        self._key_to_select = nil
-      end
-      if ImGui.BeginTabItem(ctx, tab.label, false, item_flags) then
+      if ImGui.BeginTabItem(ctx, tab.label) then
         Trap(function()
-          self:_set(tab.key)
+          self:set(tab.key)
         end)
         ImGui.EndTabItem(ctx)
       end
diff --git a/reascripts/ReaSpeech/source/SettingsControls.lua b/reascripts/ReaSpeech/source/SettingsControls.lua
index b32bde1..b2ea8ec 100644
--- a/reascripts/ReaSpeech/source/SettingsControls.lua
+++ b/reascripts/ReaSpeech/source/SettingsControls.lua
@@ -22,10 +22,6 @@ function SettingsControls:init()
     label = 'Font Size',
   }
 
-  self.font_size.on_set = function (_value)
-    self.plugin.app:refresh()
-  end
-
   self:init_logging()
   self:init_layout()
 end
```

**Comment by ramen** [comment]
> playing around with the tab bar again this morning and realizing that i forgot to make a note that the only visual weirdness i ended up seeing was when the font size is cranked up high and the text is clipped on the "Choose file" button for the file selector. i don't consider this a blocker for this particular pr. :)

Yeah, I think I should open a new issue to improve how things scale down in general. There were already issues, like with combo box widths, that are more apparent now with larger font sizes.

**Comment by ramen** [comment]
> after digging in a bit (really wanted to understand what was going on, still not 100% there), i ended up with something that appears to be functional.

Nice - feel free to commit those changes to this branch, and/or take over the PR if you like.

**Comment by mikeylove** [comment]
> Nice - feel free to commit those changes to this branch, and/or take over the PR if you like.

happy to push a commit if you like and think it's good or worth more of your time testing to feel confident about it.

asking to avoid an unrevert/more work for you in case this is an instance of "works fine for me, moving on -- oh wait sorry lol...imgui!!1" 😂

**Comment by ramen** [comment]
> > Nice - feel free to commit those changes to this branch, and/or take over the PR if you like.
> 
> happy to push a commit if you like and think it's good or worth more of your time testing to feel confident about it.
> 
> asking to avoid an unrevert/more work for you in case this is an instance of "works fine for me, moving on -- oh wait sorry lol...imgui!!1" 😂

Go ahead and push it, and I'll test on various platforms. It's easy enough to revert if it doesn't work out. Thanks!

**Comment by ramen** [comment]
I found one edge case - if the context is invalidated, we would have attempted to detach the font from the new context and not the old one, which triggers an error: "ImGui_Detach: the object is not attached to this context". I worked around this by keeping track of the context, and added pointer checks to make sure the detach is a valid operation.

File: reascripts/ReaSpeech/source/Fonts.lua [hunk 1]
```diff
@@ -5,15 +5,72 @@
...
+
+function Fonts:_detach(font)
+  if not ImGui.ValidatePtr(font, 'ImGui_Font*') then return end
+
+  local attached_ctx = self._attached_ctx[font]
+  self._attached_ctx[font] = nil
+  if not ImGui.ValidatePtr(attached_ctx, 'ImGui_Context*') then return end
+
+  ImGui.Detach(attached_ctx, font)
 end
```
Comment (mikeylove) [inline reascripts/ReaSpeech/source/Fonts.lua:74]: slick 😎
